Allow BackgroundWorkspace to be given as an OutputName

cailafinn opened the pull request:
**Description of work**

**Purpose of work**
There was a request to allow the output name from a previous row to be given as the BackgroundWorkspace on another row. The Output Name is not often the name of the output workspace as additional information is appended to the end.
**Report to:** @smk78 

**Summary of work**
This code uses the output from the normal reduction to determine what the appended suffix would be for the background workspace. It then looks for that workspace if the given BackgroundWorkspace name could not be found in the ADS.

**Further detail of work**
Giving an arbitrary workspace was part of the design requirements for this feature, so the check for the workspace with the suffix will only be done if the given name is not present in the ADS. 
Code has also been added to avoid re-running the scaling if a scaled workspace has already been created. 


**To test:**
1. Boot up workbench and the ISIS SANS interface. 
2. Load the user file and the batch file from the LOQ Demo in the TrainingCourseData
4. Process All (this will produce your "background" workspaces, something like
5. Check the `Scaled Background Subtraction` checkbox
6. Copy and paste a new row and set `Background Workspace` to `first_time`, and `Scale Factor` to 0.9.
7. Process this new row. 
8. Check the two output workspaces you should now have. If it has worked properly, the subtracted one's values should be 10% of the unsubtracted one's. 
9. Check that the batch can be saved and loaded and the background workspace and scale factor cells are preserved between loads. 

Fixes #35953 


*This does not require release notes* because **it adjusts a feature new to this release.**


<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

Commit:
Allow selection of background workspace by output name

The background workspace that is generated will generally have a suffix
applied giving the wavelength range and the reduction type info. In
order to use the GUI more effectively, add a way of supplying just the
value given in the Output Name column

RE #35953

Commit:
Fix existing tests

RE #35953

Commit:
Add tests

Yes, I should have written these as part of the development process for
the function. No, I didn't do that.

RE #35953

Commit:
Don't re-run the scaling if it already exists

RE #35953

Commit:
Update docs with new feature

RE #35953

rbauststfc reviewed (commented):
I've tested the functionality using a normal and time-sliced reduction and it's working well. Looking up the background workspace from the ADS is successful either when an exact match is provided by the user, or if matching on just an output name/sample scatter run number prefix. When the workspace for the background subtraction can't be found, a sensible error message is provided. Appropriate tests have been added.

I've made a few very minor suggestions that could be considered, but this is looking really good. 👍

When testing the time-slicing, I noticed a couple of edge cases/perhaps completely unrealistic things that I don't think need to be addressed in this PR at all, but that I thought I'd note in case there's anything we hadn't already confirmed with users, or that we might like to clarify in the documentation at some point:

- If you're performing a reduction that slices the data into multiple slices, the background subtraction will subtract the same background workspace from each slice, which I believe has been confirmed as correct behaviour with our scientists. If the user has specified the exact name of the background workspace they want to use in the interface then this all seems intuitive. If you just specify a prefix, though, then it might be less intuitive that the system will automatically choose just the first slice to use as the background rather than matching up the slices for the subtraction. I'm conscious it may be that the relevant science here means it would never be important for users to know about this, though.

- This seems likely to be an unrealistic case, or an incorrect thing to try and do, but if the background workspace data hasn't been time-sliced, then looking it up by prefix won't work if trying to apply it to a reduction that is being time-sliced. In this case, the exact name of the background workspace would need to be provided.

Commit:
Tidy docs from review and scientist feedback

RE #35953

Commit:
Avoid repeated calls to generate function

RE #35953

Commit:
Improve comment clarity

RE #35953

rbauststfc reviewed (commented):
Thanks for this - the documentation and code changes look good. I've just noticed, though, that if I try to perform the background subtraction using a valid output prefix but with the reduction mode set to `All` then I get an error message about not being able to find the workspace, rather than that the reduction mode `All` is incompatible with background subtraction.

Commit:
Fix not found error rather than all error appearing

The error to inform the user that we did not find the correct workspace
in the ADS was appearing when the reduction was set to "All" mode. This
is due to the check for the All setting being later in the reduction
before we try looking at all of the values. Move this check earlier so
that we can catch it at the right time and present the right error to
the user.

RE #35953

rbauststfc reviewed (approved):
All outstanding issues have now been addressed and this is working well.
